health/checks/checks.go

```diff
@@ -33,3 +43,115 @@ func HTTPChecker(r string) health.Checker {
 		return nil
 	})
 }
+
+// PingChecker sends an ICMP Echo request and verifies that a
+// response was reveived, and within the maximum round-trip time,
+// taking the application out of rotation otherwise
+func PingChecker(a string, maxRTT time.Duration) health.Checker {
+	return health.CheckFunc(func() error {
+		ip := net.ParseIP(a)
+		if ip == nil {
+			return fmt.Errorf("%s is not a valid IP address", a)
+		}
+		b := ip.To4()
+		pingFunc := pingIPv4
+		if b == nil {
+			pingFunc = pingIPv6
+		}
+		rtt, err := pingFunc(ip)
+		if err != nil {
+			return fmt.Errorf("ping request unsuccessful: %v", err)
+		}
+		if rtt > maxRTT {
+			return fmt.Errorf("maximum roud trip time exceeded: %v", rtt)
+		}
+		return nil
+	})
+}
+
+func pingIPv4(ip net.IP) (time.Duration, error) {
+	c, err := icmp.ListenPacket("udp4", "0.0.0.0")
+	if err != nil {
+		return 0, err
+	}
+	defer c.Close()
+
+	wm := icmp.Message{
+		Type: ipv4.ICMPTypeEcho, Code: 0,
+		Body: &icmp.Echo{
+			ID: os.Getpid() & 0xffff, Seq: 1,
+			Data: []byte("I CAN HAZ REPLY?"),
+		},
+	}
+	wb, err := wm.Marshal(nil)
+	if err != nil {
+		return 0, err
+	}
+	if _, err := c.WriteTo(wb, &net.UDPAddr{IP: ip}); err != nil {
+		return 0, err
+	}
+
+	rb := make([]byte, 1500)
+	start := time.Now()
+	deadline := start.Add(timeout)
+	c.SetReadDeadline(deadline)
```

Comment:
Nope they are distinctly different. `timeout` is the deadline on the channel so that we aren't waiting forever for an ICMP Echo Reply. The `maxRTT` is a user specified time in msec that specifies the maximum amount of time that should have elapsed between sending a response and receiving a reply. It should always be less than `timeout`